File: app/src/main/java/com/krystianwsul/checkme/DataDiff.java
package com.krystianwsul.checkme;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.krystianwsul.checkme.gui.instances.tree.GroupListFragment;
import com.krystianwsul.checkme.utils.InstanceKey;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

public class DataDiff {
    @NonNull
    private static List<String> sDiff = new ArrayList<>();

    public static void diffData(@NonNull GroupListFragment.DataWrapper oldData, @NonNull GroupListFragment.DataWrapper newData) {
        sDiff = new ArrayList<>();

        diffMap("", oldData.getInstanceDatas(), newData.getInstanceDatas());
    }

    private static void diffMap(@NonNull String indent, @NonNull Map<InstanceKey, GroupListFragment.InstanceData> oldInstanceDatas, @NonNull Map<InstanceKey, GroupListFragment.InstanceData> newInstanceDatas) {
        HashSet<InstanceKey> instanceKeys = new HashSet<>();
        instanceKeys.addAll(oldInstanceDatas.keySet());
        instanceKeys.addAll(newInstanceDatas.keySet());

        for (InstanceKey instanceKey : instanceKeys) {
            if (!oldInstanceDatas.keySet().contains(instanceKey)) {
                sDiff.add(indent + newInstanceDatas.get(instanceKey).getName() + " missing from oldData");
                continue;
            }

            GroupListFragment.InstanceData oldInstanceData = oldInstanceDatas.get(instanceKey);
            Assert.assertTrue(oldInstanceData != null);

            if (!newInstanceDatas.keySet().contains(instanceKey)) {
                sDiff.add(indent + oldInstanceData.getName() + " missing from newData");
                continue;
            }

            GroupListFragment.InstanceData newInstanceData = newInstanceDatas.get(instanceKey);
            Assert.assertTrue(newInstanceData != null);

            if (!oldInstanceData.equals(newInstanceData)) {
                sDiff.add(indent + "difference in " + oldInstanceData.getName() + ":");

                diffInstance(indent + "\t", oldInstanceData, newInstanceData);
            }
        }
    }

    private static void diffInstance(@NonNull String indent, @NonNull GroupListFragment.InstanceData oldInstanceData, @NonNull GroupListFragment.InstanceData newInstanceData) {
        Assert.assertTrue(!TextUtils.isEmpty(indent));
        Assert.assertTrue(oldInstanceData.getInstanceKey().equals(newInstanceData.getInstanceKey()));

        if (((oldInstanceData.getDone() == null) != (newInstanceData.getDone() == null)) || ((oldInstanceData.getDone() != null) && !oldInstanceData.getDone().equals(newInstanceData.getDone())))
            sDiff.add(indent + "oldInstanceData.done == " + oldInstanceData.getDone() + ", newInstanceData.done == " + newInstanceData.getDone());

        if ((TextUtils.isEmpty(oldInstanceData.getDisplayText()) != TextUtils.isEmpty(newInstanceData.getDisplayText())) || (!TextUtils.isEmpty(oldInstanceData.getDisplayText()) && !oldInstanceData.getDisplayText().equals(newInstanceData.getDisplayText())))
            sDiff.add(indent + "oldInstanceData.displayText == " + oldInstanceData.getDisplayText() + ", newInstanceData.displayText == " + newInstanceData.getDisplayText());

        if (!oldInstanceData.getName().equals(newInstanceData.getName()))
            sDiff.add(indent + "oldInstanceData.name == " + oldInstanceData.getName() + ", newInstanceData.name == " + newInstanceData.getName());

        if (!oldInstanceData.getInstanceTimeStamp().equals(newInstanceData.getInstanceTimeStamp()))
            sDiff.add(indent + "oldInstanceData.InstanceTimeStamp == " + oldInstanceData.getInstanceTimeStamp() + ", newInstanceData.InstanceTimeStamp == " + newInstanceData.getInstanceTimeStamp());

        if (oldInstanceData.getTaskCurrent() != newInstanceData.getTaskCurrent())
            sDiff.add(indent + "oldInstanceData.taskCurrent == " + oldInstanceData.getTaskCurrent() + ", newInstanceData.taskCurrent == " + newInstanceData.getTaskCurrent());

        if (oldInstanceData.getIsRootInstance() != newInstanceData.getIsRootInstance())
            sDiff.add(indent + "oldInstanceData.isRootInstance == " + oldInstanceData.getIsRootInstance() + ", newInstanceData.isRootInstance == " + newInstanceData.getIsRootInstance());

        if (((oldInstanceData.getIsRootTask() == null) != (newInstanceData.getIsRootTask() == null)) || ((oldInstanceData.getIsRootTask() != null) && !oldInstanceData.getIsRootTask().equals(newInstanceData.getIsRootTask())))
            sDiff.add(indent + "oldInstanceData.IsRootTask == " + oldInstanceData.getIsRootTask() + ", newInstanceData.IsRootTask == " + newInstanceData.getIsRootTask());

        if (oldInstanceData.getExists() != newInstanceData.getExists())
            sDiff.add(indent + "oldInstanceData.Exists == " + oldInstanceData.getExists() + ", newInstanceData.exists == " + newInstanceData.getExists());

        if (!oldInstanceData.getInstanceTimePair().equals(newInstanceData.getInstanceTimePair()))
            sDiff.add(indent + "oldInstanceData.instanceTimePair == " + oldInstanceData.getInstanceTimePair() + ", newInstanceData.instanceTimePair == " + newInstanceData.getInstanceTimePair());

        if (!oldInstanceData.getChildren().equals(newInstanceData.getChildren())) {
            sDiff.add(indent + "oldInstanceData.children != newInstanceData.children");
            diffMap(indent + "\t", oldInstanceData.getChildren(), newInstanceData.getChildren());
        }
    }

    @Nullable
    public static String getDiff() {
        return TextUtils.join("\n", sDiff);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/DataDiff.kt
package com.krystianwsul.checkme

import android.util.Log
import com.krystianwsul.checkme.gui.instances.tree.GroupListFragment
import com.krystianwsul.checkme.utils.InstanceKey
import java.util.*

object DataDiff {

    private var sDiff: MutableList<String> = ArrayList()

    val diff get() = sDiff.joinToString("\n")

    fun diffData(oldData: GroupListFragment.DataWrapper, newData: GroupListFragment.DataWrapper) {
        sDiff = ArrayList()

        diffMap("", oldData.instanceDatas, newData.instanceDatas)

        Log.e("asdf", "diff:\n$diff")
    }

    private fun diffMap(indent: String, oldInstanceDatas: Map<InstanceKey, GroupListFragment.InstanceData>, newInstanceDatas: Map<InstanceKey, GroupListFragment.InstanceData>) {
        val instanceKeys = HashSet<InstanceKey>()
        instanceKeys.addAll(oldInstanceDatas.keys)
        instanceKeys.addAll(newInstanceDatas.keys)

        for (instanceKey in instanceKeys) {
            if (!oldInstanceDatas.keys.contains(instanceKey)) {
                sDiff.add(indent + newInstanceDatas.getValue(instanceKey).name + " missing from oldData")
                continue
            }

            val oldInstanceData = oldInstanceDatas[instanceKey]
            checkNotNull(oldInstanceData)

            if (!newInstanceDatas.keys.contains(instanceKey)) {
                sDiff.add(indent + oldInstanceData.name + " missing from newData")
                continue
            }

            val newInstanceData = newInstanceDatas[instanceKey]
            checkNotNull(newInstanceData)

            if (oldInstanceData != newInstanceData) {
                sDiff.add(indent + "difference in " + oldInstanceData.name + ":")

                diffInstance(indent + "\t", oldInstanceData, newInstanceData)
            }
        }
    }

    private fun diffInstance(indent: String, oldInstanceData: GroupListFragment.InstanceData, newInstanceData: GroupListFragment.InstanceData) {
        check(indent.isNotEmpty())
        check(oldInstanceData.instanceKey == newInstanceData.instanceKey)

        if (oldInstanceData.done != newInstanceData.done)
            sDiff.add(indent + "oldInstanceData.done == " + oldInstanceData.done + ", newInstanceData.done == " + newInstanceData.done)

        if (oldInstanceData.displayText != newInstanceData.displayText)
            sDiff.add(indent + "oldInstanceData.displayText == " + oldInstanceData.displayText + ", newInstanceData.displayText == " + newInstanceData.displayText)

        if (oldInstanceData.name != newInstanceData.name)
            sDiff.add(indent + "oldInstanceData.name == " + oldInstanceData.name + ", newInstanceData.name == " + newInstanceData.name)

        if (oldInstanceData.instanceTimeStamp != newInstanceData.instanceTimeStamp)
            sDiff.add(indent + "oldInstanceData.instanceTimeStamp == " + oldInstanceData.instanceTimeStamp + ", newInstanceData.instanceTimeStamp == " + newInstanceData.instanceTimeStamp)

        if (oldInstanceData.taskCurrent != newInstanceData.taskCurrent)
            sDiff.add(indent + "oldInstanceData.taskCurrent == " + oldInstanceData.taskCurrent + ", newInstanceData.taskCurrent == " + newInstanceData.taskCurrent)

        if (oldInstanceData.isRootInstance != newInstanceData.isRootInstance)
            sDiff.add(indent + "oldInstanceData.isRootInstance == " + oldInstanceData.isRootInstance + ", newInstanceData.isRootInstance == " + newInstanceData.isRootInstance)

        if (oldInstanceData.isRootTask == null != (newInstanceData.isRootTask == null) || oldInstanceData.isRootTask != null && oldInstanceData.isRootTask != newInstanceData.isRootTask)
            sDiff.add(indent + "oldInstanceData.isRootTask == " + oldInstanceData.isRootTask + ", newInstanceData.isRootTask == " + newInstanceData.isRootTask)

        if (oldInstanceData.exists != newInstanceData.exists)
            sDiff.add(indent + "oldInstanceData.exists == " + oldInstanceData.exists + ", newInstanceData.exists == " + newInstanceData.exists)

        if (oldInstanceData.instanceTimePair != newInstanceData.instanceTimePair)
            sDiff.add(indent + "oldInstanceData.instanceTimePair == " + oldInstanceData.instanceTimePair + ", newInstanceData.instanceTimePair == " + newInstanceData.instanceTimePair)

        if (oldInstanceData.note != newInstanceData.note)
            sDiff.add(indent + "oldInstanceData.note == " + oldInstanceData.note + ", newInstanceData.note == " + newInstanceData.note)

        if (oldInstanceData.children != newInstanceData.children) {
            sDiff.add(indent + "oldInstanceData.children != newInstanceData.children")
            diffMap(indent + "\t", oldInstanceData.children, newInstanceData.children)
        }

        if (oldInstanceData.hierarchyData != newInstanceData.hierarchyData)
            sDiff.add(indent + "oldInstanceData.hierarchyData == " + oldInstanceData.hierarchyData + ", newInstanceData.hierarchyData == " + newInstanceData.hierarchyData)

        if (oldInstanceData.ordinal != newInstanceData.ordinal)
            sDiff.add(indent + "oldInstanceData.ordinal == " + oldInstanceData.ordinal + ", newInstanceData.ordinal == " + newInstanceData.ordinal)
    }
}
